proofs.go

```diff
@@ -119,37 +126,154 @@ func (proof *SparseCompactMerkleProof) Unmarshal(bz []byte) error {
+// Unmarshal deserialises the SparseMerkleClosestProof from bytes
+func (proof *SparseMerkleClosestProof) Unmarshal(bz []byte) error {
+	buf := bytes.NewBuffer(bz)
+	dec := gob.NewDecoder(buf)
+	return dec.Decode(proof)
+}
 
-		// Compact proofs: check that the correct number of sidenodes have been
-		// supplied according to the bit mask.
-		(proof.NumSideNodes > 0 && len(proof.SideNodes) != proof.NumSideNodes-countSetBits(proof.BitMask)) {
-		return false
+func (proof *SparseMerkleClosestProof) validateBasic(spec *TreeSpec) error {
+	// ensure the depth of the leaf node being proven is within the path size
+	if proof.Depth < 0 || proof.Depth > spec.ph.PathSize()*8 {
+		return fmt.Errorf("invalid depth: %d", proof.Depth)
 	}
+	// for each of the bits flipped ensure that they are within the path size
+	// and that they are not greater than the depth of the leaf node being proven
+	for _, i := range proof.FlippedBits {
+		// as proof.Depth <= spec.ph.PathSize()*8, i <= proof.Depth
+		if i < 0 || i > proof.Depth {
+			return fmt.Errorf("invalid flipped bit index: %d", i)
+		}
+	}
+	// create the path of the leaf node using the flipped bits metadata
+	workingPath := proof.Path
+	for _, i := range proof.FlippedBits {
+		flipPathBit(workingPath, i)
+	}
+	// ensure that the path of the leaf node being proven has a prefix
+	// of length depth as the path provided (with bits flipped)
+	if equal, failed := equalPrefixBits(
+		workingPath,
+		proof.ClosestPath,
+		0, proof.Depth,
+	); !equal {
+		return fmt.Errorf("invalid closest path: %x (not equal at bit: %d)", proof.ClosestPath, failed)
+	}
+	// validate the proof itself
+	if err := proof.ClosestProof.validateBasic(spec); err != nil {
+		return fmt.Errorf("invalid closest proof: %w", err)
+	}
+	return nil
+}
 
-	return true
+// SparseCompactMerkleClosestProof is a compressed representation of the SparseMerkleClosestProof
+type SparseCompactMerkleClosestProof struct {
+	Path             []byte                    // the path provided to the ProveClosest method
+	FlippedBits      [][]byte                  // the index of the bits flipped in the path during tree traversal
+	Depth            []byte                    // the depth of the tree when tree traversal stopped
+	ClosestPath      []byte                    // the path of the leaf closest to the path provided
+	ClosestValueHash []byte                    // the value hash of the leaf closest to the path provided
+	ClosestProof     *SparseCompactMerkleProof // the proof of the leaf closest to the path provided
+}
+
+func (proof *SparseCompactMerkleClosestProof) validateBasic(spec *TreeSpec) error {
+	// Do a basic sanity check on the proof on the fields of the proof specific to
+	// the compact proof only.
+	//
+	// When the proof is de-compacted and verified, the sanity check for the
+	// de-compacted proof should be executed.
+
+	// ensure no compressed fields are larger than the path size
+	maxSliceLen := minBytes(spec.ph.PathSize() * 8)
```

Comment:
```suggestion
	// for example, for a 256-bit hasher, minBytes will return 1 and require
	// all downstream values to have a length of at most one byte
	maxSliceLen := minBytes(spec.ph.PathSize() * 8)
```

Reply:
You've already done the thinking - just document it to save time for someone in the future